### PR TITLE
Friends list improvements (including contact groups and activity sort)

### LINKED ISSUE #1125: Contact grouping / categorization
One should be able to sort their contacts into various categories.

```
Work
* contact
* contact
School
* contact
* contact
Friends
* contact
* contact
```

### PR DESCRIPTION
**Pull request by TheSpiritXIII**
This is a big commit, meant to improve consistency and introduce new features. It includes the following changes:
- Organize friends using “circles”. Includes a compact mode.
- Search and filter through friends within circles.
- Ability to cycle between circles. Cycling is also now O(log n) instead of O(n).
- Friends add/status change find fixed from O(n) to O(log n).
- Dragging friends over groups or circles highlights the group or circle.
- Inline renaming of friends, groups and circles.
- When groups and circles are added, users rename them immediately.
- Groups and circles are sorted alphabetically.
- Notification expand the circle that the friend is in.
- View friends by sorting by activity.

Tested extensively with Valgrind. Let me know if there are any problems.

This fixes: #1125 #1356

### CODE REVIEW

**Comment by agilob** [comment]
gj, screenshot?

File: qtox.pro [hunk 1]
```diff
@@ -517,4 +522,9 @@ HEADERS += \
     src/core/toxid.h \
     src/persistence/profile.h \
     src/widget/translator.h \
-    src/persistence/settingsserializer.h
+    src/persistence/settingsserializer.h \
+    src/widget/genericchatitemwidget.h \
+    src/widget/circlewidget.h \
+    src/widget/friendlistlayout.h \
+	src/widget/genericchatitemlayout.h \
```
Comment (zetok) [inline qtox.pro]: > tab

spaces pls

File: src/widget/friendlistwidget.cpp [hunk 2]
```diff
@@ -16,122 +16,552 @@
...
+    today = today.addDays(-6);
+    if (date >= today)
+        return ThisWeek;
+
+    today = today.addDays(-today.day() + 1); // Go to the beginning of the month.
+    if (last7DaysWasLastMonth())
+    {
+        if (date >= today)
+            return ThisMonth;
+        today = today.addMonths(-1);
```
Comment (zetok) [inline src/widget/friendlistwidget.cpp:83]: should be a newline between

```
return ThisMonth;
```

and

```
today = today.addMonths(-1);
```

File: src/widget/friendlistwidget.cpp [hunk 3]
```diff
@@ -16,122 +16,552 @@
...
+            continue;
+        }
 
-            friends.append(widget);
+        // Go to the actual next index.
+        if (currentLayout == listLayout->getLayoutOnline() || currentLayout == listLayout->getLayoutOffline() || currentLayout == groupLayout.getLayout())
+        {
+            GenericChatroomWidget* chatWidget = dynamic_cast<GenericChatroomWidget*>(currentLayout->itemAt(index)->widget());
+            if (chatWidget != nullptr)
+                emit chatWidget->chatroomWidgetClicked(chatWidget);
```
Comment (zetok) [inline src/widget/friendlistwidget.cpp:558]: `\n`

File: src/widget/friendlistwidget.cpp [hunk 4]
```diff
@@ -141,3 +571,80 @@ void FriendListWidget::reDraw()
...
+}
+
+QLayout* FriendListWidget::nextLayout(QLayout* layout, bool forward) const
+{
+    if (layout == groupLayout.getLayout())
+    {
+        if (forward)
+        {
+            if (groupsOnTop)
+                return listLayout->getLayoutOnline();
```
Comment (zetok) [inline src/widget/friendlistwidget.cpp:685]: `\n`

File: src/widget/friendlistwidget.cpp [hunk 5]
```diff
@@ -141,3 +571,80 @@ void FriendListWidget::reDraw()
...
+            return listLayout->getLayoutOffline();
+        }
+        else
+        {
+            if (groupsOnTop)
+                return circleLayout->getLayout();
```
Comment (zetok) [inline src/widget/friendlistwidget.cpp:692]: `\n`

File: src/widget/friendlistwidget.cpp [hunk 6]
```diff
@@ -141,3 +571,80 @@ void FriendListWidget::reDraw()
...
+            return listLayout->getLayoutOnline();
+        }
+    }
+    else if (layout == listLayout->getLayoutOnline())
+    {
+        if (forward)
+        {
+            if (groupsOnTop)
+                return listLayout->getLayoutOffline();
```
Comment (zetok) [inline src/widget/friendlistwidget.cpp:702]: `\n`

File: src/widget/friendlistwidget.cpp [hunk 7]
```diff
@@ -141,3 +571,80 @@ void FriendListWidget::reDraw()
...
+            return groupLayout.getLayout();
+        }
+        else
+        {
+            if (groupsOnTop)
+                return groupLayout.getLayout();
```
Comment (zetok) [inline src/widget/friendlistwidget.cpp:709]: `\n`

File: src/widget/friendlistwidget.cpp [hunk 8]
```diff
@@ -141,3 +571,80 @@ void FriendListWidget::reDraw()
...
+            return circleLayout->getLayout();
+        }
+    }
+    else if (layout == listLayout->getLayoutOffline())
+    {
+        if (forward)
+            return circleLayout->getLayout();
+        else if (groupsOnTop)
+            return listLayout->getLayoutOnline();
```
Comment (zetok) [inline src/widget/friendlistwidget.cpp:719]: `\n`

File: src/widget/friendlistwidget.cpp [hunk 9]
```diff
@@ -141,3 +571,80 @@ void FriendListWidget::reDraw()
...
+        return groupLayout.getLayout();
+    }
+    else if (layout == circleLayout->getLayout())
+    {
+        if (forward)
+        {
+            if (groupsOnTop)
+                return groupLayout.getLayout();
```
Comment (zetok) [inline src/widget/friendlistwidget.cpp:728]: `\n`

File: src/widget/friendwidget.cpp [hunk 10]
```diff
@@ -48,28 +51,65 @@ FriendWidget::FriendWidget(int FriendId, QString id)
...
+    if (circleWidget == nullptr)
+        friendList = dynamic_cast<FriendListWidget*>(parentWidget());
+    else
+        friendList = dynamic_cast<FriendListWidget*>(circleWidget->parentWidget());
+
+        circleMenu = menu.addMenu(tr("Move to circle...", "Menu to move a friend into a different circle"));
+
+        newCircleAction = circleMenu->addAction(tr("To new circle"));
+        if (circleId != -1)
+            removeCircleAction = circleMenu->addAction(tr("Remove from circle '%1'").arg(Settings::getInstance().getCircleName(circleId)));
```
Comment (zetok) [inline src/widget/friendwidget.cpp]: `\n`

File: src/widget/genericchatitemlayout.cpp [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+
+    int index = indexOfClosestSortedWidget(widget);
+    if (index >= layout->count())
+        return -1;
+
+    GenericChatItemWidget* atMid = dynamic_cast<GenericChatItemWidget*>(layout->itemAt(index)->widget());
+    assert(atMid != nullptr);
+
+    if (atMid == widget)
+        return index;
```
Comment (zetok) [inline src/widget/genericchatitemlayout.cpp:56]: `\n`

File: src/widget/genericchatitemlayout.cpp [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+
+bool GenericChatItemLayout::existsSortedWidget(GenericChatItemWidget* widget) const
+{
+    return indexOfSortedWidget(widget) != -1;
+}
+
+void GenericChatItemLayout::removeSortedWidget(GenericChatItemWidget* widget)
+{
+    if (layout->isEmpty())
+        return;
```
Comment (zetok) [inline src/widget/genericchatitemlayout.cpp:69]: `\n`

**Comment by TheSpiritXIII** [comment]
@agilob Sure
screenie
screenie
I'm wondering whether the white text for the new selector works (where you select either by name or by activity). The old combobox has gray text for me.

**Comment by agilob** [comment]
Nice, thanks. 

// please ignore zetok.

**Comment by zetok** [comment]
max width panel
↑ panel gets maximized for no reason, can't be minimized.

**Comment by TheSpiritXIII** [comment]
Everything was fixed. Let me know if you find anything else.

**Comment by zetok** [comment]
searching is buggy in all ways
↑ searching is buggy when one switches between activity/name. (yes, there should be contacts there)

**Comment by ProMcTagonist** [comment]
spacing corrections

1:  There should not be a scrollbar trough here when there's no scrollbar. Don't know if that's from this PR or not.
2 & 3:  There shouldn't be a space here, it makes it look floaty and cluttered. Much more noticeable in 2 than in 3, but it should be fixed uniformly.

**Comment by zetok** [comment]
Horribly buggy.
1. Even though I had conversations with contacts, they're all grouped under "Unknown" "group" when sorting by activity
2. When joining groupchat, its title is being changed…

**Comment by zetok** [comment]
https://gist.github.com/zetok/96f43544444917fcb2a2 ← segfault when switching sorting `active ↔ by name`

**Comment by zetok** [comment]
Also, it looks like my status message was changed to `Your status` when I've used qTox with this patch applied…

**Comment by zetok** [comment]
There is a bug when changing layout of contact list from `compact` → `normal`:
muh full size
that being said…
@ProMcTagonist ↑ would that be a proper size for ~~groups~~ err, circles?

**Comment by TheSpiritXIII** [comment]
@ProMcTagonist How small should they be? Here they are 10 pixels smaller;

And here, 10 pixels smaller than that:

**Comment by zetok** [comment]
1. search for someone
2. make circles while text is stil entered in search field ↓



~~On a plus side, now (so far) I wasn't able to make it crash~~ actually, it crashed

**Comment by zetok** [comment]
Latest crash, compiled with `-O0`
backtrace: https://gist.github.com/zetok/96f43544444917fcb2a2#file-gistfile3-txt
`t a a bt f`: https://gist.github.com/zetok/96f43544444917fcb2a2#file-gistfile4-txt

**Comment by zetok** [comment]
Crash when removing empty circles: https://gist.github.com/zetok/20b239a34b49716ec043

**Comment by zetok** [comment]
https://d.maxfile.ro/njvpyvxhhx.webm ← issues with renaming

**Comment by ProMcTagonist** [comment]
@TheSpiritXIII They should be touching.

**Comment by TheSpiritXIII** [comment]
@ProMcTagonist In that case, should we just use compact mode design for regular mode as well as [zetok suggested](https://github.com/tux3/qTox/pull/1879#issuecomment-113141375)?

**Comment by zetok** [comment]
https://gist.github.com/zetok/4fbf7014892db9444f75 ← different segfault

**Comment by ProMcTagonist** [comment]
I just want to avoid the floating white line. I can't apply a diff right now so I don't know exactly how it'll look but it should be a step up at least. Sorry I missed your suggestion, @zetok, thanks :)

**Comment by zetok** [comment]
moar sort
↑ sorting circles still doesn't work quite right in menu on 1dd42f5295c6babcc0340736e93addb980b25ea3

Edit: actually, I've just noticed that there are apparently circles that have duplicated names `-.-`

**Comment by suhr** [comment]
btw it should use [natural sort](http://blog.codinghorror.com/sorting-for-humans-natural-sort-order/)

**Comment by zetok** [comment]
There is still bug present with groupchat being renamed

**Comment by zetok** [comment]
Actually, groupchat is not being really renamed, but real title is not being displayed:
real title pls

**Comment by TheSpiritXIII** [comment]
@suhr It does - it uses the QCollator class which includes natural sorting. It was my first time using it so maybe there was a setting I missed.

**Comment by TheSpiritXIII** [comment]
So, let's talk behavior for the activity sort. I talked to @tux3 about it.

Here's how it currently works: It saves all activity in the settings file. If the friend you are looking at is not in the file, it queries the history database. Any further activity will be stored in the settings file. I made this descision so that people who do not have the history saving enabled will still be able to use the activity sort. The issue with this is that it is slow.

What @tux3 is thinking is that those who do not have the history saving enabled should **not** be able to use the activity sorting. Rather, they will be able to sort by activity but **only the activity that occured while they had qTox open - their activity is lost when they close it**. This would be a lot faster (as all of the querying could be done through the chat form instead of a database or settings file). However, I'm not sure if this is what users would expect.

So: **Do people who do not save their history want have their activity saved or not?**

**Comment by zetok** [comment]
> What @tux3 is thinking is that those who do not have the history saving enabled should **not** be able to use the activity sorting. Rather, they will be able to sort by activity but **only the activity that occured while they had qTox open - their activity is lost when they close it**.

Yep. Exactly how it should be done.

**Comment by ProMcTagonist** [comment]
People who don't save history emphatically do not want their activity saved.

**Comment by zetok** [comment]
@TheSpiritXIII same renaming issue with `i3-4.10.2`

**Comment by TheSpiritXIII** [comment]
Another thing to dicuss: Circles are currently placed at the bottom. @zetok says that this makes them "annoying to use". Any suggestions?

**Comment by ProMcTagonist** [comment]
Are circles folders or tags?

**Comment by TheSpiritXIII** [comment]
@ProMcTagonist What's the difference? You can only have a single level of circles, so they're not really folders (if traditional folders are trees).

**Comment by ProMcTagonist** [comment]
Can a friend be in multiple circles?

**Comment by zetok** [comment]
@ProMcTagonist nope, they're are 1-level "folders"

**Comment by tux3** [comment]
The background of a circle seems to be ignoring the theme color after a restart 
ss011

**Comment by zetok** [comment]
@tux3 ok, your branch `friendlist` @ ddd6f741d90b8be23a2f2cc337104e330b4a248e improved situation a bit:

Partially fixed:
- [ ] multiple simultaneous renames (it's harder to trigger now, but still possible)

Not fixed: 
- [ ] Name being editable e.g. after I minimized qTox to tray and restored it from tray
- [ ] Renaming not working if some other input field already has "focus"

**Comment by tux3** [comment]
Those bugs aren't in qTox, so this is ready for merging anyways.

I rebased it on master in branch friendlist.

**Pull request closed without merging**